apm-lambda-extension/extension/apm_server.go

```diff
@@ -48,30 +44,85 @@ const (
+	config            *extensionConfig
+	AgentDoneSignal   chan struct{}
+	dataChannel       chan AgentData
+	client            *http.Client
+	status            ApmServerTransportStatusType
+	reconnectionCount int
+	gracePeriodTimer  *time.Timer
 }
 
-// The status of transport to the APM server.
-//
-// This instance of the ApmServerTransportStateType is public for use in tests.
-var ApmServerTransportState = ApmServerTransportStateType{
-	Status:            Healthy,
-	ReconnectionCount: -1,
+func InitApmServerTransport(config *extensionConfig) *ApmServerTransport {
+	var transport ApmServerTransport
+	transport.bufferPool = sync.Pool{New: func() interface{} {
+		return &bytes.Buffer{}
+	}}
+	transport.AgentDoneSignal = make(chan struct{}, 1)
+	transport.dataChannel = make(chan AgentData, 100)
+	transport.client = &http.Client{
+		Timeout:   time.Duration(config.DataForwarderTimeoutSeconds) * time.Second,
+		Transport: http.DefaultTransport.(*http.Transport).Clone(),
+	}
+	transport.config = config
+	transport.status = Healthy
+	transport.reconnectionCount = -1
+	return &transport
+}
+
+// StartBackgroundApmDataForwarding Receive agent data as it comes in and post it to the APM server.
+// Stop checking for, and sending agent data when the function invocation
+// has completed, signaled via a channel.
+func (transport *ApmServerTransport) ForwardApmData(ctx context.Context) error {
+	if transport.status == Failing {
+		return nil
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			Log.Debug("Invocation context cancelled, not processing any more agent data")
+			return nil
+		case agentData := <-transport.dataChannel:
+			if err := PostToApmServer(ctx, transport, agentData); err != nil {
+				return fmt.Errorf("error sending to APM server, skipping: %v", err)
+			}
+		}
+	}
+}
+
+// FlushAPMData reads all the apm data in the apm data channel and sends it to the APM server.
+func FlushAPMData(ctx context.Context, transport *ApmServerTransport) {
+	if transport.status == Failing {
+		Log.Debug("Flush skipped - Transport failing")
+		return
+	}
+	Log.Debug("Flush started - Checking for agent data")
+	for {
+		select {
+		case agentData := <-transport.dataChannel:
+			Log.Debug("Flush in progress - Processing agent data")
+			if err := PostToApmServer(ctx, transport, agentData); err != nil {
+				Log.Errorf("Error sending to APM server, skipping: %v", err)
+			}
+		default:
+			Log.Debug("Flush ended - No agent data on buffer")
+			return
+		}
+	}
 }
 
 // PostToApmServer takes a chunk of APM agent data and posts it to the APM server.
 //
 // The function compresses the APM agent data, if it's not already compressed.
 // It sets the APM transport status to failing upon errors, as part of the backoff
 // strategy.
-func PostToApmServer(client *http.Client, agentData AgentData, config *extensionConfig, ctx context.Context) error {
+func PostToApmServer(ctx context.Context, transport *ApmServerTransport, agentData AgentData) error {
```

Comment:
```suggestion
func (transport *ApmServerTransport) PostToApmServer(ctx context.Context, agentData AgentData) error {
```

Reply:
Implemented in https://github.com/elastic/apm-aws-lambda/commit/d71ccfb98277e9d98fd558d9fd3be9cf796a6515